File: backend/src/test/java/lan/dk/podcastserver/entity/PodcastTest.java
package lan.dk.podcastserver.entity;

import io.vavr.collection.HashSet;
import io.vavr.control.Option;
import org.junit.Before;
import org.junit.Test;
import org.springframework.util.FileSystemUtils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.ZonedDateTime;
import java.util.UUID;

import static lan.dk.podcastserver.assertion.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * Created by kevin on 15/06/15 for HackerRank problem
 */
public class PodcastTest {

    private static final ZonedDateTime NOW = ZonedDateTime.now();
    private static final Cover COVER = Cover.builder().url("ACover.jpg").build();
    private static String PODCAST_TO_STRING;
    private Podcast podcast;
    private UUID id;

    @Before
    public void init() {
        Podcast.rootFolder = Paths.get("/tmp");

        /* Given */
        id = UUID.randomUUID();
        podcast = Podcast
                .builder()
                    .id(id)
                    .title("PodcastDeTest")
                    .url("http://nowhere.com")
                    .signature("ae4b93a7e8249d6be591649c936dbe7d")
                    .type("Youtube")
                    .lastUpdate(NOW)
                    .cover(COVER)
                    .description("A long Description")
                    .hasToBeDeleted(true)
                    .items(HashSet.<Item>empty().toJavaSet())
                    .tags(HashSet.<Tag>empty().toJavaSet())
                .build();

        PODCAST_TO_STRING = "Podcast{id="+ id +", title='PodcastDeTest', url='http://nowhere.com', signature='ae4b93a7e8249d6be591649c936dbe7d', type='Youtube', lastUpdate=%s}";

        FileSystemUtils.deleteRecursively(Podcast.rootFolder.resolve(podcast.getTitle()).toFile());
    }

    @Test
    public void should_have_all_setters_and_getters_working() {
        /* Then */
        assertThat(podcast)
                .hasId(id)
                .hasTitle("PodcastDeTest")
                .hasUrl("http://nowhere.com")
                .hasSignature("ae4b93a7e8249d6be591649c936dbe7d")
                .hasType("Youtube")
                .hasLastUpdate(NOW)
                .hasCover(COVER)
                .hasDescription("A long Description")
                .hasHasToBeDeleted(true)
                .hasNoItems()
                .hasNoTags();
    }
    
    @Test
    public void should_have_toString() {
        assertThat(podcast.toString()).isEqualTo(String.format(PODCAST_TO_STRING, NOW));
    }
    
    @Test
    public void should_have_hashcode_and_equals() {
        /* Given */
        Podcast samePodcast = new Podcast();
        samePodcast.setId(id);
        samePodcast.setLastUpdate(NOW);
        samePodcast.setSignature("ae4b93a7e8249d6be591649c936dbe7d");
        samePodcast.setTitle("PodcastDeTest");
        samePodcast.setUrl("http://nowhere.com");

        Object notPodcast = new Object();

        /* Then */
        assertThat(podcast).isEqualTo(podcast);
        assertThat(podcast).isNotEqualTo(notPodcast);
        assertThat(podcast).isEqualTo(samePodcast);
        assertThat(podcast.hashCode()).isNotNull();
    }
    
    @Test
    public void should_add_an_item() {
        Item itemToAdd = new Item();
        /* When */
        podcast.add(itemToAdd);
        /* Then */
        assertThat(itemToAdd).hasPodcast(podcast);
        assertThat(podcast).hasOnlyItems(itemToAdd);
    }

    @Test
    public void should_contains_item() {
        Item itemToAdd = new Item().setId(UUID.randomUUID());
        Item itemToAdd2 = new Item().setId(UUID.randomUUID());
        podcast.add(itemToAdd);

        assertThat(podcast.contains(itemToAdd)).isTrue();
        assertThat(podcast.contains(itemToAdd2)).isFalse();
    }
    
    @Test
    public void should_be_update_now() {
        /* When */
        podcast.lastUpdateToNow();
        /* Then */
        assertThat(podcast.getLastUpdate().isAfter(NOW));
    }

    @Test
    public void should_delete_on_post_remove() throws IOException {
        /* Given */
        Podcast.rootFolder = Paths.get("/tmp");
        podcast.setHasToBeDeleted(true);
        Path podcastFolder = Podcast.rootFolder.resolve("PodcastDeTest");
        Path coverFile = podcastFolder.resolve("cover.jpg");
        Files.createDirectory(podcastFolder);
        Files.createFile(coverFile);

        /* When */
        podcast.postRemove();

        /* Then */
        assertThat(coverFile).doesNotExist();
        assertThat(podcastFolder).doesNotExist();
    }

    @Test
    public void should_not_delete_if_podcast_is_not_auto_delete() throws IOException {
        Podcast.rootFolder = Paths.get("/tmp");
        podcast.setHasToBeDeleted(false);
        Path podcastFolder = Podcast.rootFolder.resolve("PodcastDeTest");
        Path coverFile = podcastFolder.resolve("cover.jpg");
        Files.createDirectory(podcastFolder);
        Files.createFile(coverFile);

        /* When */
        podcast.postRemove();

        /* Then */
        assertThat(coverFile).exists();
        assertThat(podcastFolder).exists();
    }

    @Test
    public void should_provide_cover_path() {
        /* GIVEN */
        /* WHEN  */
        Option<Path> coverPath = podcast.getCoverPath();

        /* THEN  */
        assertThat(coverPath).contains(Podcast.rootFolder.resolve(podcast.getTitle()).resolve("cover.jpg"));
    }

}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/entity/PodcastTest.kt
package com.github.davinkevin.podcastserver.entity

import io.vavr.collection.HashSet
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.util.FileSystemUtils
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.time.ZonedDateTime
import java.util.UUID

/**
 * Created by kevin on 15/06/15 for HackerRank problem
 */
class PodcastTest {
    private lateinit var podcast: Podcast
    private lateinit var anId: UUID

    @BeforeEach
    fun init() {
        Podcast.rootFolder = Paths.get("/tmp")

        /* Given */
        anId = UUID.randomUUID()
        podcast = Podcast().apply {
            id = anId
            title = "PodcastDeTest"
            url = "http://nowhere.com"
            signature = "ae4b93a7e8249d6be591649c936dbe7d"
            type = "Youtube"
            lastUpdate = NOW
            cover = COVER
            description = "A long Description"
            hasToBeDeleted = true
            items = HashSet.empty<Item>().toJavaSet()
            tags = HashSet.empty<Tag>().toJavaSet()
        }

        PODCAST_TO_STRING = "Podcast{id=$anId, title='PodcastDeTest', url='http://nowhere.com', signature='ae4b93a7e8249d6be591649c936dbe7d', type='Youtube', lastUpdate=%s}"

        FileSystemUtils.deleteRecursively(Podcast.rootFolder!!.resolve(podcast.title).toFile())
    }

    @Test
    fun should_have_all_setters_and_getters_working() {
        /* Then */
        assertThat(podcast.id).isEqualTo(anId)
        assertThat(podcast.title).isEqualTo("PodcastDeTest")
        assertThat(podcast.url).isEqualTo("http://nowhere.com")
        assertThat(podcast.signature).isEqualTo("ae4b93a7e8249d6be591649c936dbe7d")
        assertThat(podcast.type).isEqualTo("Youtube")
        assertThat(podcast.lastUpdate).isEqualTo(NOW)
        assertThat(podcast.cover).isEqualTo(COVER)
        assertThat(podcast.description).isEqualTo("A long Description")
        assertThat(podcast.hasToBeDeleted).isEqualTo(true)
        assertThat(podcast.items).isEmpty()
        assertThat(podcast.tags).isEmpty()
    }

    @Test
    fun should_have_toString() {
        assertThat(podcast.toString())
                .isEqualTo(String.format(PODCAST_TO_STRING, NOW))
    }

    @Test
    fun should_have_hashcode_and_equals() {
        /* Given */
        val samePodcast = Podcast()
        samePodcast.id = anId
        samePodcast.lastUpdate = NOW
        samePodcast.signature = "ae4b93a7e8249d6be591649c936dbe7d"
        samePodcast.title = "PodcastDeTest"
        samePodcast.url = "http://nowhere.com"

        val notPodcast = Any()

        /* Then */
        assertThat(podcast).isEqualTo(podcast)
        assertThat(podcast).isNotEqualTo(notPodcast)
        assertThat(podcast).isEqualTo(samePodcast)
        assertThat(podcast.hashCode()).isNotNull()
    }

    @Test
    fun should_add_an_item() {
        val itemToAdd = Item()
        /* When */
        podcast.add(itemToAdd)
        /* Then */
        assertThat(itemToAdd.podcast).isEqualTo(podcast)
        assertThat(podcast.items).containsOnly(itemToAdd)
    }

    @Test
    fun should_contains_item() {
        val itemToAdd = Item().apply { id = UUID.randomUUID() }
        val itemToAdd2 = Item().apply { id = UUID.randomUUID() }
        podcast.add(itemToAdd)

        assertThat(podcast.contains(itemToAdd)).isTrue()
        assertThat(podcast.contains(itemToAdd2)).isFalse()
    }


    @Test
    fun should_provide_cover_path() {
        /* GIVEN */
        /* WHEN  */
        val coverPath = podcast.coverPath

        /* THEN  */
        assertThat(coverPath).contains(Podcast.rootFolder!!.resolve(podcast.title).resolve("cover.jpg"))
    }

    companion object {

        private val NOW = ZonedDateTime.now()
        private val COVER = object : Cover() {
            init {
                url = "ACover.jpg"
            }
        }
        private lateinit var PODCAST_TO_STRING: String
    }

}
